File: agents/base_agent.py
"""
DEPRECATED: This Python agent is being migrated to Go.

Per the project's language policy, all agents must be implemented in:
- C++, DLLs, GO, Rust, or R

Python is no longer permitted for agent implementations.

Migration Status:
- This file will be replaced by base_agent.go
- See agents/LANGUAGE_POLICY.md for details

Original: Base Agent Class for Auto-Improvement System
"""

import asyncio
import json
import logging
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List, Optional, Union
from enum import Enum

import structlog
from pydantic import BaseModel, Field

# Configure logging
structlog.configure(
    processors=[
        structlog.processors.TimeStamper(fmt="iso"),
        structlog.processors.JSONRenderer()
    ],
    wrapper_class=structlog.BoundLogger,
    context_class=dict,
    logger_factory=structlog.PrintLoggerFactory(),
)

logger = structlog.get_logger()


class AgentState(Enum):
    """Possible states for an agent."""
    IDLE = "idle"
    RUNNING = "running"
    PAUSED = "paused"
    ERROR = "error"
    COMPLETED = "completed"


class TaskResult(BaseModel):
    """Result of an agent task execution."""
    success: bool = True
    data: Dict[str, Any] = Field(default_factory=dict)
    errors: List[str] = Field(default_factory=list)
    metrics: Dict[str, float] = Field(default_factory=dict)
    timestamp: datetime = Field(default_factory=datetime.now)


class AgentMemory(BaseModel):
    """Memory system for storing agent state and learning."""
    short_term: Dict[str, Any] = Field(default_factory=dict)
    long_term: Dict[str, Any] = Field(default_factory=dict)
    patterns: List[Dict[str, Any]] = Field(default_factory=list)
    improvements: List[Dict[str, Any]] = Field(default_factory=list)
    
    def store_pattern(self, pattern: Dict[str, Any]) -> None:
        """Store a detected pattern."""
        pattern["timestamp"] = datetime.now().isoformat()
        self.patterns.append(pattern)
        # Keep only last 100 patterns
        if len(self.patterns) > 100:
            self.patterns = self.patterns[-100:]
    
    def store_improvement(self, improvement: Dict[str, Any]) -> None:
        """Store an implemented improvement."""
        improvement["timestamp"] = datetime.now().isoformat()
        self.improvements.append(improvement)
        # Keep only last 50 improvements
        if len(self.improvements) > 50:
            self.improvements = self.improvements[-50:]
    
    def get_recent_patterns(self, limit: int = 10) -> List[Dict[str, Any]]:
        """Get recent patterns."""
        return self.patterns[-limit:]
    
    def get_recent_improvements(self, limit: int = 10) -> List[Dict[str, Any]]:
        """Get recent improvements."""
        return self.improvements[-limit:]


@dataclass
class AgentConfig:
    """Configuration for an agent."""
    name: str
    description: str
    system_prompt: str
    tools: List[str] = field(default_factory=list)
    max_iterations: int = 5
    improvement_cycle_frequency: str = "daily"
    rate_limit_rpm: int = 15  # Gemini free tier
    rate_limit_tpm: int = 1_000_000
    cache_ttl: int = 3600


class BaseAgent(ABC):
    """
    Base agent class providing common functionality.
    
    All specialized agents (Research, Analysis, Code, Test) inherit from this class.
    """
    
    def __init__(self, config: AgentConfig, llm_service=None):
        self.config = config
        self.llm_service = llm_service
        self.state = AgentState.IDLE
        self.memory = AgentMemory()
        self.logger = logger.bind(agent=config.name)
        self._start_time: Optional[datetime] = None
        self._task_count: int = 0
        
        # Rate limiting state
        self._request_timestamps: List[datetime] = []
        self._token_usage: int = 0
        self._daily_requests: int = 0
        
        # Performance tracking
        self._performance_metrics: Dict[str, List[float]] = {
            "response_time": [],
            "token_usage": [],
            "success_rate": [],
        }
    
    @property
    def name(self) -> str:
        return self.config.name
    
    @property
    def description(self) -> str:
        return self.config.description
    
    @property
    def uptime(self) -> Optional[float]:
        """Get agent uptime in seconds."""
        if self._start_time:
            return (datetime.now() - self._start_time).total_seconds()
        return None
    
    async def initialize(self) -> None:
        """Initialize the agent."""
        self.logger.info("Initializing agent", agent=self.name)
        self.state = AgentState.IDLE
        self._start_time = datetime.now()
        await self._load_memory()
        self.logger.info("Agent initialized", agent=self.name)
    
    async def shutdown(self) -> None:
        """Shutdown the agent gracefully."""
        self.logger.info("Shutting down agent", agent=self.name)
        self.state = AgentState.IDLE
        await self._save_memory()
        self.logger.info("Agent shut down", agent=self.name)
    
    async def _load_memory(self) -> None:
        """Load agent memory from persistent storage."""
        memory_path = Path(f"agents/memory/{self.name.lower()}_memory.json")
        if memory_path.exists():
            try:
                data = json.loads(memory_path.read_text(encoding="utf-8"))
                self.memory = AgentMemory(**data)
                self.logger.info("Memory loaded", patterns=len(self.memory.patterns))
            except Exception as e:
                self.logger.error("Failed to load memory", error=str(e))
    
    async def _save_memory(self) -> None:
        """Save agent memory to persistent storage."""
        memory_path = Path(f"agents/memory/{self.name.lower()}_memory.json")
        memory_path.parent.mkdir(parents=True, exist_ok=True)
        try:
            memory_path.write_text(
                self.memory.model_dump_json(indent=2),
                encoding="utf-8"
            )
            self.logger.info("Memory saved", path=str(memory_path))
        except Exception as e:
            self.logger.error("Failed to save memory", error=str(e))
    
    def check_rate_limit(self, estimated_tokens: int = 1000) -> bool:
        """
        Check if request is within rate limits.
        
        Args:
            estimated_tokens: Estimated token usage for the request
            
        Returns:
            True if request is allowed, False otherwise
        """
        now = datetime.now()
        
        # Clean old timestamps (older than 1 minute)
        one_minute_ago = now.timestamp() - 60
        self._request_timestamps = [
            ts for ts in self._request_timestamps
            if ts.timestamp() > one_minute_ago
        ]
        
        # Check RPM limit
        if len(self._request_timestamps) >= self.config.rate_limit_rpm:
            self.logger.warning("RPM limit reached", 
                              current=len(self._request_timestamps),
                              limit=self.config.rate_limit_rpm)
            return False
        
        # Check TPM limit
        if self._token_usage + estimated_tokens > self.config.rate_limit_tpm:
            self.logger.warning("TPM limit approaching",
                              current=self._token_usage,
                              limit=self.config.rate_limit_tpm)
            return False
        
        return True
    
    def record_request(self, tokens_used: int) -> None:
        """Record a request for rate limiting."""
        self._request_timestamps.append(datetime.now())
        self._token_usage += tokens_used
        self._daily_requests += 1
    
    def get_performance_summary(self) -> Dict[str, Any]:
        """Get performance summary for the agent."""
        return {
            "name": self.name,
            "state": self.state.value,
            "uptime_seconds": self.uptime,
            "total_tasks": self._task_count,
            "daily_requests": self._daily_requests,
            "token_usage": self._token_usage,
            "patterns_stored": len(self.memory.patterns),
            "improvements_stored": len(self.memory.improvements),
            "avg_response_time": (
                sum(self._performance_metrics["response_time"]) / 
                len(self._performance_metrics["response_time"])
                if self._performance_metrics["response_time"] else 0
            ),
        }
    
    async def execute_with_retry(
        self,
        task_func,
        max_retries: int = 3,
        delay: float = 1.0,
        *args,
        **kwargs
    ) -> TaskResult:
        """
        Execute a task with retry logic.
        
        Args:
            task_func: Async function to execute
            max_retries: Maximum number of retry attempts
            delay: Delay between retries in seconds
            *args, **kwargs: Arguments to pass to task_func
            
        Returns:
            TaskResult with execution results
        """
        last_error = None
        
        for attempt in range(max_retries):
            try:
                self.logger.info("Executing task", attempt=attempt + 1)
                result = await task_func(*args, **kwargs)
                self._task_count += 1
                return result
            except Exception as e:
                last_error = str(e)
                self.logger.error("Task execution failed",
                                attempt=attempt + 1,
                                error=last_error)
                if attempt < max_retries - 1:
                    await asyncio.sleep(delay * (attempt + 1))
        
        return TaskResult(
            success=False,
            errors=[f"Task failed after {max_retries} attempts: {last_error}"]
        )
    
    @abstractmethod
    async def execute_task(self, task: Dict[str, Any]) -> TaskResult:
        """
        Execute a task. Must be implemented by subclasses.
        
        Args:
            task: Task specification
            
        Returns:
            TaskResult with execution results
        """
        pass
    
    @abstractmethod
    async def analyze_performance(self) -> Dict[str, Any]:
        """
        Analyze agent performance. Must be implemented by subclasses.
        
        Returns:
            Dictionary with performance analysis
        """
        pass
    
    @abstractmethod
    async def suggest_improvements(self) -> List[Dict[str, Any]]:
        """
        Suggest improvements based on analysis. Must be implemented by subclasses.
        
        Returns:
            List of improvement suggestions
        """
        pass

File: agents/pkg/base_agent.go
package pkg

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// AgentState represents possible states for an agent
type AgentState string

const (
	AgentStateIdle      AgentState = "idle"
	AgentStateRunning   AgentState = "running"
	AgentStatePaused    AgentState = "paused"
	AgentStateError     AgentState = "error"
	AgentStateCompleted AgentState = "completed"
)

// TaskResult represents result of an agent task execution
type TaskResult struct {
	SuccessInt int8                   `json:"success"` // 0/1 instead of bool
	Data       map[string]interface{} `json:"data"`
	Errors     []string               `json:"errors"`
	Metrics    map[string]int64       `json:"metrics"` // Fixed-point (scale 1000 = 1.0)
	Timestamp  time.Time              `json:"timestamp"`
}

// NewTaskResult creates a new TaskResult with default values
func NewTaskResult() *TaskResult {
	return &TaskResult{
		SuccessInt: 1, // true = 1
		Data:       make(map[string]interface{}),
		Errors:     make([]string, 0),
		Metrics:    make(map[string]int64),
		Timestamp:  time.Now(),
	}
}

// AgentMemory stores agent state and learning data
type AgentMemory struct {
	ShortTerm    map[string]interface{}   `json:"short_term"`
	LongTerm     map[string]interface{}   `json:"long_term"`
	Patterns     []map[string]interface{} `json:"patterns"`
	Improvements []map[string]interface{} `json:"improvements"`
	mu           sync.RWMutex
}

// NewAgentMemory creates a new AgentMemory with default values
func NewAgentMemory() *AgentMemory {
	return &AgentMemory{
		ShortTerm:    make(map[string]interface{}),
		LongTerm:     make(map[string]interface{}),
		Patterns:     make([]map[string]interface{}, 0),
		Improvements: make([]map[string]interface{}, 0),
	}
}

// StorePattern stores a detected pattern
func (m *AgentMemory) StorePattern(pattern map[string]interface{}) {
	m.mu.Lock()
	defer m.mu.Unlock()

	pattern["timestamp"] = time.Now().Format(time.RFC3339)
	m.Patterns = append(m.Patterns, pattern)

	// Keep only last 100 patterns
	if len(m.Patterns) > 100 {
		m.Patterns = m.Patterns[len(m.Patterns)-100:]
	}
}

// StoreImprovement stores an implemented improvement
func (m *AgentMemory) StoreImprovement(improvement map[string]interface{}) {
	m.mu.Lock()
	defer m.mu.Unlock()

	improvement["timestamp"] = time.Now().Format(time.RFC3339)
	m.Improvements = append(m.Improvements, improvement)

	// Keep only last 50 improvements
	if len(m.Improvements) > 50 {
		m.Improvements = m.Improvements[len(m.Improvements)-50:]
	}
}

// GetRecentPatterns returns recent patterns
func (m *AgentMemory) GetRecentPatterns(limit int) []map[string]interface{} {
	m.mu.RLock()
	defer m.mu.RUnlock()

	if limit <= 0 || limit > len(m.Patterns) {
		limit = len(m.Patterns)
	}
	return m.Patterns[len(m.Patterns)-limit:]
}

// GetRecentImprovements returns recent improvements
func (m *AgentMemory) GetRecentImprovements(limit int) []map[string]interface{} {
	m.mu.RLock()
	defer m.mu.RUnlock()

	if limit <= 0 || limit > len(m.Improvements) {
		limit = len(m.Improvements)
	}
	return m.Improvements[len(m.Improvements)-limit:]
}

// AgentConfig contains configuration for an agent
type AgentConfig struct {
	Name                      string   `json:"name"`
	Description               string   `json:"description"`
	SystemPrompt              string   `json:"system_prompt"`
	Tools                     []string `json:"tools"`
	MaxIterations             int      `json:"max_iterations"`
	ImprovementCycleFrequency string   `json:"improvement_cycle_frequency"`
	RateLimitRPM              int      `json:"rate_limit_rpm"`
	RateLimitTPM              int      `json:"rate_limit_tpm"`
	CacheTTL                  int      `json:"cache_ttl"`
}

// BaseAgent is the base agent class providing common functionality
type BaseAgent struct {
	Config    *AgentConfig
	State     AgentState
	Memory    *AgentMemory
	StartTime time.Time
	TaskCount int
	mu        sync.Mutex

	// Rate limiting state
	requestTimestamps []time.Time
	tokenUsage        int
	dailyRequests     int

	// Performance tracking (fixed-point)
	performanceMetrics map[string][]int64
}

// NewBaseAgent creates a new BaseAgent
func NewBaseAgent(config *AgentConfig) *BaseAgent {
	// Set defaults
	if config.RateLimitRPM == 0 {
		config.RateLimitRPM = 15 // Gemini free tier
	}
	if config.RateLimitTPM == 0 {
		config.RateLimitTPM = 1_000_000
	}
	if config.MaxIterations == 0 {
		config.MaxIterations = 5
	}

	return &BaseAgent{
		Config:             config,
		State:              AgentStateIdle,
		Memory:             NewAgentMemory(),
		performanceMetrics: make(map[string][]int64),
	}
}

// Name returns the agent name
func (a *BaseAgent) Name() string {
	return a.Config.Name
}

// Uptime returns agent uptime in seconds
func (a *BaseAgent) Uptime() float64 {
	if a.StartTime.IsZero() {
		return 0
	}
	return time.Since(a.StartTime).Seconds()
}

// Initialize initializes the agent
func (a *BaseAgent) Initialize() error {
	a.mu.Lock()
	defer a.mu.Unlock()

	a.State = AgentStateIdle
	a.StartTime = time.Now()

	if err := a.loadMemory(); err != nil {
		return fmt.Errorf("failed to load memory: %w", err)
	}

	return nil
}

// Shutdown shuts down the agent gracefully
func (a *BaseAgent) Shutdown() error {
	a.mu.Lock()
	defer a.mu.Unlock()

	a.State = AgentStateIdle

	if err := a.saveMemory(); err != nil {
		return fmt.Errorf("failed to save memory: %w", err)
	}

	return nil
}

// loadMemory loads agent memory from persistent storage
func (a *BaseAgent) loadMemory() error {
	memoryPath := filepath.Join("agents", "memory", fmt.Sprintf("%s_memory.json", a.Name()))

	if _, err := os.Stat(memoryPath); os.IsNotExist(err) {
		// Memory file doesn't exist yet, that's fine
		return nil
	}

	data, err := os.ReadFile(memoryPath)
	if err != nil {
		return err
	}

	return json.Unmarshal(data, a.Memory)
}

// saveMemory saves agent memory to persistent storage
func (a *BaseAgent) saveMemory() error {
	memoryDir := filepath.Join("agents", "memory")
	if err := os.MkdirAll(memoryDir, 0755); err != nil {
		return err
	}

	memoryPath := filepath.Join(memoryDir, fmt.Sprintf("%s_memory.json", a.Name()))

	data, err := json.MarshalIndent(a.Memory, "", "  ")
	if err != nil {
		return err
	}

	return os.WriteFile(memoryPath, data, 0644)
}

// CheckRateLimit checks if request is within rate limits
// Returns int8: 1 = allowed, 0 = denied (GF(3) compliant)
func (a *BaseAgent) CheckRateLimit(estimatedTokens int) int8 {
	a.mu.Lock()
	defer a.mu.Unlock()

	now := time.Now()
	oneMinuteAgo := now.Add(-60 * time.Second)

	// Clean old timestamps
	filtered := make([]time.Time, 0, len(a.requestTimestamps))
	for _, ts := range a.requestTimestamps {
		if ts.After(oneMinuteAgo) {
			filtered = append(filtered, ts)
		}
	}
	a.requestTimestamps = filtered

	// Check RPM limit
	if len(a.requestTimestamps) >= a.Config.RateLimitRPM {
		return 0 // false = 0
	}

	// Check TPM limit
	if a.tokenUsage+estimatedTokens > a.Config.RateLimitTPM {
		return 0 // false = 0
	}

	return 1 // true = 1
}

// RecordRequest records a request for rate limiting
func (a *BaseAgent) RecordRequest(tokensUsed int) {
	a.mu.Lock()
	defer a.mu.Unlock()

	a.requestTimestamps = append(a.requestTimestamps, time.Now())
	a.tokenUsage += tokensUsed
	a.dailyRequests++
}

// GetPerformanceSummary returns performance summary for the agent
func (a *BaseAgent) GetPerformanceSummary() map[string]interface{} {
	a.mu.RLock()
	defer a.mu.RUnlock()

	summary := map[string]interface{}{
		"name":                a.Name(),
		"state":               string(a.State),
		"uptime_seconds":      a.Uptime(),
		"total_tasks":         a.TaskCount,
		"daily_requests":      a.dailyRequests,
		"token_usage":         a.tokenUsage,
		"patterns_stored":     len(a.Memory.Patterns),
		"improvements_stored": len(a.Memory.Improvements),
	}

	// Calculate average response time (fixed-point)
	rt, ok := a.performanceMetrics["response_time"]
	if ok && len(rt) > 0 {
		sum := int64(0)
		for _, t := range rt {
			sum += t
		}
		summary["avg_response_time"] = sum / int64(len(rt))
	} else {
		summary["avg_response_time"] = 0
	}

	return summary
}

// ExecuteWithRetry executes a task with retry logic
func (a *BaseAgent) ExecuteWithRetry(taskFunc func() (*TaskResult, error), maxRetries int, delay time.Duration) (*TaskResult, error) {
	if maxRetries <= 0 {
		maxRetries = 3
	}
	if delay <= 0 {
		delay = 1 * time.Second
	}

	var lastError error

	for attempt := 0; attempt < maxRetries; attempt++ {
		result, err := taskFunc()
		if err == nil {
			a.mu.Lock()
			a.TaskCount++
			a.mu.Unlock()
			return result, nil
		}

		lastError = err

		if attempt < maxRetries-1 {
			time.Sleep(delay * time.Duration(attempt+1))
		}
	}

	return nil, fmt.Errorf("task failed after %d attempts: %w", maxRetries, lastError)
}

// Agent interface that all specialized agents must implement
type Agent interface {
	Initialize() error
	Shutdown() error
	ExecuteTask(task map[string]interface{}) (*TaskResult, error)
	AnalyzePerformance() (map[string]interface{}, error)
	SuggestImprovements() ([]map[string]interface{}, error)
}
